### PR TITLE
complete code of customizing devise

### PR DESCRIPTION
**Pull request by alix-jhonson**
This is a code of customizing devise.

### CODE REVIEW

File: app/views/devise/sessions/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<% title "Sign in" %>
+
+<%= form_for(resource_name, resource, :url => session_path(resource_name)) do |f| %>
```
Comment (SumraYasmin) [inline app/views/devise/sessions/new.html.erb]: use new syntax
`url: session_path(resource_name)`

File: app/views/devise/sessions/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  <p><%= f.label :username %><br />
+  <%= f.text_field :username %></p>
+
+  <p><%= f.label :password %><br />
+  <%= f.password_field :password %></p>
+
+  <!-- ... -->
```
Comment (SumraYasmin) [inline app/views/devise/sessions/new.html.erb]: submit button is missing?

File: config/initializers/devise.rb [hunk 3]
```diff
@@ -268,6 +268,9 @@
   # The default HTTP method used to sign out a resource. Default is :delete.
   config.sign_out_via = :delete
 
+  config.authentication_keys = [ :username ]
+config.password_length = 4..20
```
Comment (SumraYasmin) [inline config/initializers/devise.rb]: indentation issue

File: config/initializers/devise.rb [hunk 4]
```diff
@@ -268,6 +268,9 @@
   # The default HTTP method used to sign out a resource. Default is :delete.
   config.sign_out_via = :delete
 
+  config.authentication_keys = [ :username ]
```
Comment (SumraYasmin) [inline config/initializers/devise.rb]: no spaces inside `[]`
`[:username ]`

File: app/controllers/projects_controller.rb [hunk 5]
```diff
@@ -0,0 +1 @@
+before_filter :authenticate_user!, :except => [:show, :index]
```
Comment (SumraYasmin) [inline app/controllers/projects_controller.rb:1]: Do we need to authenticate before show and index ?

File: app/views/devise/sessions/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<% title "Sign in" %>
+
+<%= form_for(resource_name, resource, :url => session_path(resource_name)) do |f| %>
+  <p><%= f.label :username %><br />
+  <%= f.text_field :username %></p>
```
Comment (SumraYasmin) [inline app/views/devise/sessions/new.html.erb:5]: Do we need to permit `username` in devise params?

**Review by SumraYasmin — CHANGES_REQUESTED** [review]

File: app/controllers/projects_controller.rb (hunk 5 above)
Comment (alix-jhonson) [reply app/controllers/projects_controller.rb:1]: Here, In this code, this helper is restricting access to the other actions except for show and index.

File: app/views/devise/sessions/new.html.erb (hunk 1 above)
Comment (alix-jhonson) [reply app/views/devise/sessions/new.html.erb]: done.

Comment (alix-jhonson) [reply app/views/devise/sessions/new.html.erb]: done

File: app/views/devise/sessions/new.html.erb (hunk 2 above)
Comment (alix-jhonson) [reply app/views/devise/sessions/new.html.erb]: done

File: config/initializers/devise.rb (hunk 3 above)
Comment (alix-jhonson) [reply config/initializers/devise.rb]: done

File: config/initializers/devise.rb (hunk 4 above)
Comment (alix-jhonson) [reply config/initializers/devise.rb]: done.

File: app/views/devise/sessions/new.html.erb (hunk 6 above)
Comment (alix-jhonson) [reply app/views/devise/sessions/new.html.erb:5]: done.

**Pull request merged**